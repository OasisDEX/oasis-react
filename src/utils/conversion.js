import BigNumber from "bignumber.js";
import tokens from "../store/selectors/tokens";
import { APP_BASE_PRECISION } from "../constants";

let getState = null;

export function convertToTokenPrecision(amountIn18Precision, tokenName) {
  const tokenSpecs = tokens.getTokenSpecs(getState(), tokenName);
  if (tokenSpecs.get("precision") === APP_BASE_PRECISION) {
    return amountIn18Precision;
  } else {
    const precisionDifference =
      APP_BASE_PRECISION - tokenSpecs.get("precision");
    return new BigNumber(amountIn18Precision)
      .times(new BigNumber(10).pow(-precisionDifference))
      .toFixed(0);
  }
}

export function convertTo18Precision(amountInTokenPrecision, token) {
  const tokenSpecs = tokens.getTokenSpecs(getState(), token);
  if (tokenSpecs.get("precision") === APP_BASE_PRECISION) {
    return amountInTokenPrecision;
  } else {
    return new BigNumber(amountInTokenPrecision)
      .times(
        new BigNumber(10).pow(APP_BASE_PRECISION - tokenSpecs.get("precision"))
      )
      .toFixed(0);
  }
}

const init = getStateFunction => (getState = getStateFunction);

export default {
  init
};
