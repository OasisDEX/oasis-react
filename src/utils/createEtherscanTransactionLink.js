import { KOVAN, MAIN } from '../constants';

const createEtherscanTransactionLink = ({ activeNetworkName, transactionHash }) => {
  switch (activeNetworkName) {
    case MAIN: return `https://etherscan.io/tx/${transactionHash}`;
    case KOVAN: return `https://${activeNetworkName}.etherscan.io/tx/${transactionHash}`
  }
};


export default createEtherscanTransactionLink;