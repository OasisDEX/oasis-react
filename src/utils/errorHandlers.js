import { ERRORS } from './errorCodes';

import { onMissingConnection } from './onMissingConnection';

const handlers = {
  [ERRORS.NO_CONNECTION]: onMissingConnection,
};

class ErrorHandler {
  handle(error) {
    console.log('ErrorHandler', error);
    const handler = handlers[error];
    if (!handler) {
      throw new Error('Platform error:', error);
    }

    handler(error);
  }
}

export const errorHandler = new ErrorHandler();