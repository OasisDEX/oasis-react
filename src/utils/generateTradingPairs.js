import { TOKEN_MAKER } from '../constants';

export const generateTradingPairs = (baseTokens, quoteTokens) => {
  const TradingPairs = [
    {
      base: 'MKR',
      quote: 'W-ETH',
      priority: 10,
      isDefault: true
    },
    {
      base: 'W-ETH',
      quote: 'DAI',
      priority: 9,
      isDefault: true
    },
    {
      base: 'MKR',
      quote: 'DAI',
      priority: 8,
      isDefault: true
    },
    {
      base: 'SAI',
      quote: 'DAI',
      priority: 6,
      isDefault: true
    },
  ];

  baseTokens.forEach((base) => {
    if(base === TOKEN_MAKER) { return; }
    quoteTokens.forEach((quote) => {
      TradingPairs.push({
        base,
        quote,
        priority: 0,
        isDefault: false
      });
    });
  });
  return TradingPairs;
};