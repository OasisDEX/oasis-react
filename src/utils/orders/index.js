import { formatAmount, formatPrice } from '../tokens/pair';
import { ETH_UNIT_ETHER } from '../../constants';
import { TYPE_BUY_OFFER, TYPE_SELL_OFFER } from '../../store/reducers/offers';
import web3 from '../../bootstrap/web3';
import { getTokenContractInstance } from '../../bootstrap/contracts';


const toDisplayFormat = offer => ({

  tradeType: offer.tradeType,
  ask_price: formatPrice(offer.ask_price),
  bid_price: formatPrice(offer.bid_price),
  sell_how_much: formatAmount(offer.sellHowMuch, true,  ETH_UNIT_ETHER),
  buy_how_much: formatAmount(offer.buyHowMuch, true, ETH_UNIT_ETHER),
  ask_price_full_precision: offer.ask_price,
  bid_price_full_precision: offer.bid_price,
  sell_how_much_full_precision: offer.sellHowMuch,
  buy_how_much_full_precision: offer.buyHowMuch,
  owner: offer.owner,
  id: offer.id
});

const getOfferType = (baseToken, { sellWhichTokenAddress, buyWhichTokenAddress }) => {
  const baseTokenAddress = getTokenContractInstance(baseToken).address;
  switch (baseTokenAddress) {
    case sellWhichTokenAddress: return TYPE_SELL_OFFER;
    case buyWhichTokenAddress: return TYPE_BUY_OFFER;
  }
};

const isOfferOwner = ({ owner }, address ) => {
  if(address) {
    return (owner === address);
  } else {
    return owner === web3.eth.defaultAccount;
  }
};

export {
  toDisplayFormat,
  getOfferType,
  isOfferOwner,
}