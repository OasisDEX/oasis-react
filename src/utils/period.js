import network from '../store/selectors/network';
import platform from '../store/selectors/platform';

let getState = null;

const init = (getStateFunction) => {
  getState = getStateFunction;
};

const getNetworkMeta = () => network.activeNetworkMeta(getState());
const getActiveNetworkAvgBlocksPerDay = () => parseInt(getNetworkMeta().get('avgBlocksPerDay'));

export const isLeapYear = (year) => year % 400 === 0 || (year % 100 !== 0 && year % 4 === 0);

export const DAY = 'TIME_SPAN/DAY';
export const WEEK = 'TIME_SPAN/WEEK';
export const TWO_WEEKS = 'TIME_SPAN/TWO_WEEKS';
export const MONTH = 'TIME_SPAN/MONTH';
export const QUARTER_OF_YEAR = 'TIME_SPAN/QUARTER_OF_YEAR';
export const HALF_OF_YEAR = 'TIME_SPAN/HALF_OF_YEAR';
export const YEAR = 'TIME_SPAN/YEAR';
export const TWO_YEARS = 'TIME_SPAN/TWO_YEARS';


export const getNumberOfDaysInTheYear = (year) => isLeapYear(year) ? 366 : 365;
const avgBlockPerDefaultPeriod = () => avgBlockPer(platform.defaultPeriod(getState()));

const avgBlockPerActivePeriod = () => avgBlockPer(platform.activePeriod(getState()));

const avgBlockPer = (period) => {
  const avgBlocksPerDay = getActiveNetworkAvgBlocksPerDay();
  switch (period) {
    case DAY: return avgBlocksPerDay;
    case WEEK: return 7 * avgBlocksPerDay;
    case TWO_WEEKS: return 14 * avgBlocksPerDay;
    default: throw new Error('Wrong time span provided!');
  }
};

export default {
  init,
  avgBlockPer,
  avgBlockPerActivePeriod,
  avgBlockPerDefaultPeriod
}

