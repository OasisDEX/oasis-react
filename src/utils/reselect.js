const getProps = (...args) => args[1];

export default {
  getProps
}