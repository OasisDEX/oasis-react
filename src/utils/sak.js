import balances from '../store/reducers/balances';
import web3, {web3p} from "../bootstrap/web3";
import markets from "../store/selectors/markets";
import {ETH_UNIT_ETHER} from "../constants";
import { getMarketContractInstance } from '../bootstrap/contracts';
import BigNumber from "bignumber.js";

const fromWei = (x) =>  web3.fromWei(x, ETH_UNIT_ETHER);

const toString = (x) => x.toString();

const print = (x) => console.log(x)

const marketAddress = (store) =>
  () => markets.activeMarketAddress(store.getState());

const getTokenAllowance = (store) =>
  (tokenName) => {
    const x = balances
      .actions
      .getDefaultAccountTokenAllowanceForAddress(tokenName, marketAddress(store)())((a) => store.dispatch(a), () => store.getState());

    return x
      .then(({value}) => value)
      .then(fromWei)
      .then(toString)
      .then(print);
  }

const disableTokenTrust = (store) =>
  (tokenName) =>
    balances.actions.setTokenTrustAddressDisabled(tokenName, marketAddress(store)()).payload
      .then(watch);

const enableTokenTrust = (store) =>
  (tokenName) =>
    balances.actions.setTokenTrustAddressEnabled(tokenName, marketAddress(store)()).payload
      .then(watch);

const watch = (txHash) => {
  const filter = web3.eth.filter('latest');
  console.log(`Watching transaction: ${txHash}`);
  filter.watch(function(error) {
    if(error) {
      console.log("Watch error:", error);
      return;
    }
    web3p.eth.getTransactionReceipt(txHash)
      .then((receipt) => {
        if(receipt) {
          console.log(`Transaction: ${txHash}, status: ${receipt.status === '0x1' ? 'success' : 'failure'}`, receipt);
          filter.stopWatching();
        }
      });
  });
};

const logTake = (fromBlock, toBlock ) => {
  getMarketContractInstance().LogTake({}, {fromBlock, toBlock})
    .get((err, logTakesList) => {
      console.log('logTakesList', fromBlock, toBlock, logTakesList);
    });
};

const loadOffer = (offerId) => {
  getMarketContractInstance().offers(offerId)
    .then((o) => o.map(x => x.toString()))
    .then(print);
}

export default (store) => {
  window.sak = {
    getTokenAllowance: getTokenAllowance(store),
    disableTokenTrust: disableTokenTrust(store),
    enableTokenTrust: enableTokenTrust(store),
    state: () => store.getState().toJS(),
    logTake,
    BigNumber: (n) => new BigNumber(n),
    loadOffer
  };
}