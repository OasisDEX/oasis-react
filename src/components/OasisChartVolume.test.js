/* global describe it expect */
/* eslint-disable import/first */
import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';
import { mockDate } from '../utils/testHelpers'

import {
  OasisChartVolume,
  mapStateToProps,
  mapDispatchToProps,
} from './OasisChartVolume';

describe('(Component) OasisChartVolume', () => {
  const state = fromJS(global.storeMock);
  const initialProps = mockDate('2018-05-10', () => mapStateToProps(state));
  const initialActions = mapDispatchToProps(x => x);
  const props = {
    ...initialActions,
    ...initialProps,
    tradingPair: {
      baseToken: 'empty1',
      quoteToken: 'empty2',
    },
  };

  it('should render', () => {
    const wrapper = shallow(
      <OasisChartVolume {...props}/>,
    );
    // expect(wrapper).toMatchSnapshot();
  });
});
