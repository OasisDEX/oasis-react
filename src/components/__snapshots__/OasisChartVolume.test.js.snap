// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) OasisChartVolume should render 1`] = `
<Line
  data={
    Object {
      "datasets": Array [
        Object {
          "backgroundColor": "rgba(140, 133, 200, 0.1)",
          "borderColor": "#8D86C9",
          "borderWidth": 3,
          "data": Array [
            "0.00000",
            "0.00000",
            "0.00000",
            "0.00000",
            "0.00000",
            "0.00000",
            "0.00000",
          ],
          "label": "Volume",
          "pointBackgroundColor": "#8D86C9",
          "pointRadius": 3,
        },
      ],
      "labels": Array [
        "2018-05-04T22:00:00.000Z",
        "2018-05-05T22:00:00.000Z",
        "2018-05-06T22:00:00.000Z",
        "2018-05-07T22:00:00.000Z",
        "2018-05-08T22:00:00.000Z",
        "2018-05-09T22:00:00.000Z",
        "2018-05-10T22:00:00.000Z",
      ],
    }
  }
  options={
    Object {
      "layout": Object {
        "padding": 5,
      },
      "legend": Object {
        "display": false,
      },
      "scales": Object {
        "xAxes": Array [
          Object {
            "display": false,
          },
        ],
        "yAxes": Array [
          Object {
            "ticks": Object {
              "beginAtZero": true,
            },
          },
        ],
      },
      "tooltips": Object {
        "enabled": false,
      },
    }
  }
/>
`;
